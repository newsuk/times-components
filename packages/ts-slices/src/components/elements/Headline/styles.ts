import get from 'lodash.get';

import styled from 'styled-components';
import { breakpoints, colours, fonts } from '@times-components/styleguide';

import { DisplaySchema } from '../../../types/styles';
import { isCentered } from '../../../utils/getArticleStyles';

const getFontSize = (breakpoint: string, schema?: DisplaySchema) =>
  get(get(schema, breakpoint), 'headlineFontSize', 10);

const getWidth = (breakpoint: string, schema?: DisplaySchema) =>
  isCentered(breakpoint, schema)
    ? breakpoint === 'sm' || breakpoint === 'md'
      ? '90%'
      : '80%'
    : '100%';

const getTextAlign = (breakpoint: string, schema?: DisplaySchema) =>
  isCentered(breakpoint, schema) ? 'center' : 'left';

export const HeadlineContainer = styled.div<{ schema?: DisplaySchema }>`
  width: ${({ schema }) => getWidth('sm', schema)};
  margin: 0 auto;

  @media (min-width: ${breakpoints.medium}px) {
    width: ${({ schema }) => getWidth('md', schema)};
  }

  @media (min-width: ${breakpoints.wide}px) {
    width: ${({ schema }) => getWidth('lg', schema)};
  }

  @media (min-width: ${breakpoints.huge}px) {
    width: ${({ schema }) => getWidth('xlg', schema)};
  }

  a {
    text-decoration: none;

    h3 {
      margin: 0;
      color: ${colours.functional.brandColour};
      font-size: ${({ schema }) => getFontSize('sm', schema)}px;
      line-height: ${({ schema }) => getFontSize('sm', schema)}px;
      font-family: ${fonts.headline};
      font-weight: normal;
      text-align: ${({ schema }) => getTextAlign('sm', schema)};

      @media (min-width: ${breakpoints.medium}px) {
        font-size: ${({ schema }) => getFontSize('md', schema)}px;
        line-height: ${({ schema }) => getFontSize('md', schema)}px;
        text-align: ${({ schema }) => getTextAlign('md', schema)};
      }

      @media (min-width: ${breakpoints.wide}px) {
        font-size: ${({ schema }) => getFontSize('lg', schema)}px;
        line-height: ${({ schema }) => getFontSize('lg', schema)}px;
        text-align: ${({ schema }) => getTextAlign('lg', schema)};
      }

      @media (min-width: ${breakpoints.huge}px) {
        font-size: ${({ schema }) => getFontSize('xlg', schema)}px;
        line-height: ${({ schema }) => getFontSize('xlg', schema)}px;
        text-align: ${({ schema }) => getTextAlign('xlg', schema)};
      }
    }

    &:hover {
      h3 {
        color: #006699;
      }
    }
  }
`;
