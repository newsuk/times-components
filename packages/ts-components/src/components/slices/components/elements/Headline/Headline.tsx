import React from 'react';

import { SliceArticle } from '../../../types/slice';
import { MouseEventType, ClickHandlerType } from '../../../types/event';
import { DisplaySchema } from '../../../types/styles';

import { HeadlineContainer } from './styles';

export const Headline: React.FC<{
  article: SliceArticle;
  displaySchema?: DisplaySchema;
  clickHandler?: ClickHandlerType;
}> = ({ article, displaySchema, clickHandler }) => {
  if (!article.headline) {
    return null;
  }

  const onClick = (event: MouseEventType) => {
    if (article && clickHandler) {
      clickHandler(event, article);
    }
  };

  return (
    <HeadlineContainer schema={displaySchema}>
      <a onClick={onClick} href={article.url}>
        <h3>{article.headline}</h3>
      </a>
    </HeadlineContainer>
  );
};
