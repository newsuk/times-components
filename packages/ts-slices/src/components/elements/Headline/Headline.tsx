import React from 'react';

import { SliceArticle } from '../../../types/slice';
import { MouseEventType, ClickHandlerType } from '../../../types/event';
import { DisplaySchema } from '../../../types/styles';

import { HeadlineContainer } from './styles';

export const Headline: React.FC<{
  article: SliceArticle;
  displaySchema?: DisplaySchema;
  clickHandler?: ClickHandlerType;
}> = ({ article, displaySchema, clickHandler }) => {
  if (!article.headline) {
    return null;
  }

  const onClick = (event: MouseEventType) => {
    if (article && clickHandler) {
      clickHandler(event, article);
    }
    // location.href is required instead of <a href={} />
    // this is a side effect caused by transformChannelData
    // changing article urls client-side causes hydration warning
    location.href = article.url;
  };

  return (
    <HeadlineContainer schema={displaySchema}>
      <a onClick={onClick} href={article.url}>
        <h3>{article.headline}</h3>
      </a>
    </HeadlineContainer>
  );
};
