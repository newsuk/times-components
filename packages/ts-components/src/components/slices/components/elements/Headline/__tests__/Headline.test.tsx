import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Headline } from '../Headline';

const mockFireEvent = jest.fn();

const article = {
  images: { crops: [] },
  headline: 'Some example headline',
  url: ''
};

describe('<Headline />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should render headline component correctly', () => {
    const { asFragment, getByText } = render(<Headline article={article} />);

    getByText(article.headline);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should fire the correct event when the headline clicked', () => {
    const { getByText } = render(
      <Headline article={article} clickHandler={mockFireEvent} />
    );

    fireEvent.click(getByText(article.headline));

    expect(mockFireEvent).toHaveBeenCalled();
  });
});
