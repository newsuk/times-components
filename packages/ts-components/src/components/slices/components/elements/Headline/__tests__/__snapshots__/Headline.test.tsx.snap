// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Headline /> should render headline component correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa elQUxc"
  >
    <a
      href=""
    >
      <h3>
        Some example headline
      </h3>
    </a>
  </div>
</DocumentFragment>
`;
